File: ant-sdk/antd-go/integration_test.go
package antd

import (
	"context"
	"errors"
	"os"
	"testing"
	"time"
)

// integrationClient returns a Client pointed at a live antd daemon.
// It skips the test if the daemon is not reachable.
func integrationClient(t *testing.T) *Client {
	t.Helper()

	url := os.Getenv("ANTD_TEST_URL")
	if url == "" {
		url = "http://127.0.0.1:51105"
	}

	c := NewClient(url, WithTimeout(10*time.Second))

	// Probe health to decide whether to skip.
	ctx, cancel := context.WithTimeout(context.Background(), 5*time.Second)
	defer cancel()
	_, err := c.Health(ctx)
	if err != nil {
		t.Skipf("skipping integration test: antd daemon not reachable at %s: %v", url, err)
	}
	return c
}

func TestIntegration_Health(t *testing.T) {
	c := integrationClient(t)
	h, err := c.Health(context.Background())
	if err != nil {
		t.Fatalf("Health failed: %v", err)
	}
	if !h.OK {
		t.Fatal("expected health OK=true")
	}
	if h.Network != "local" {
		t.Fatalf("expected network=local, got %q", h.Network)
	}
}

func TestIntegration_DataPutPublic_NoWallet(t *testing.T) {
	c := integrationClient(t)
	_, err := c.DataPutPublic(context.Background(), []byte("hello"), PaymentModeAuto)
	if err == nil {
		t.Fatal("expected error from DataPutPublic without wallet")
	}
	var svcErr *ServiceUnavailableError
	var netErr *NetworkError
	if !errors.As(err, &svcErr) && !errors.As(err, &netErr) {
		t.Fatalf("expected ServiceUnavailableError or NetworkError, got %T: %v", err, err)
	}
}

func TestIntegration_DataGetPublic_InvalidAddress(t *testing.T) {
	c := integrationClient(t)
	_, err := c.DataGetPublic(context.Background(), "not-a-valid-address")
	if err == nil {
		t.Fatal("expected error for invalid address")
	}
	// Could be 400 (bad address format) or 404 (not found) or 502 (no peers).
	var badReq *BadRequestError
	var notFound *NotFoundError
	var netErr *NetworkError
	var svcErr *ServiceUnavailableError
	if !errors.As(err, &badReq) && !errors.As(err, &notFound) && !errors.As(err, &netErr) && !errors.As(err, &svcErr) {
		t.Fatalf("expected BadRequestError, NotFoundError, NetworkError, or ServiceUnavailableError, got %T: %v", err, err)
	}
}

func TestIntegration_WalletAddress_NoWallet(t *testing.T) {
	c := integrationClient(t)
	_, err := c.WalletAddress(context.Background())
	if err == nil {
		t.Fatal("expected error from WalletAddress without wallet")
	}
	var svcErr *ServiceUnavailableError
	if !errors.As(err, &svcErr) {
		t.Fatalf("expected ServiceUnavailableError, got %T: %v", err, err)
	}
}

func TestIntegration_WalletBalance_NoWallet(t *testing.T) {
	c := integrationClient(t)
	_, err := c.WalletBalance(context.Background())
	if err == nil {
		t.Fatal("expected error from WalletBalance without wallet")
	}
	var svcErr *ServiceUnavailableError
	if !errors.As(err, &svcErr) {
		t.Fatalf("expected ServiceUnavailableError, got %T: %v", err, err)
	}
}

func TestIntegration_WalletApprove_NoWallet(t *testing.T) {
	c := integrationClient(t)
	err := c.WalletApprove(context.Background())
	if err == nil {
		t.Fatal("expected error from WalletApprove without wallet")
	}
	var svcErr *ServiceUnavailableError
	if !errors.As(err, &svcErr) {
		t.Fatalf("expected ServiceUnavailableError, got %T: %v", err, err)
	}
}

func TestIntegration_FinalizeUpload_InvalidID(t *testing.T) {
	c := integrationClient(t)
	_, err := c.FinalizeUpload(context.Background(), "nonexistent-upload-id", map[string]string{"q": "t"}, false)
	if err == nil {
		t.Fatal("expected error for nonexistent upload_id")
	}
	var notFound *NotFoundError
	var badReq *BadRequestError
	var svcErr *ServiceUnavailableError
	if !errors.As(err, &notFound) && !errors.As(err, &badReq) && !errors.As(err, &svcErr) {
		t.Fatalf("expected NotFoundError, BadRequestError, or ServiceUnavailableError, got %T: %v", err, err)
	}
}

func TestIntegration_DataCost_NoPeers(t *testing.T) {
	c := integrationClient(t)
	_, err := c.DataCost(context.Background(), []byte("test data"), PaymentModeAuto)
	if err == nil {
		t.Fatal("expected error from DataCost without peers")
	}
	// No peers means 502 (NetworkError) or 503 (ServiceUnavailableError).
	var netErr *NetworkError
	var svcErr *ServiceUnavailableError
	if !errors.As(err, &netErr) && !errors.As(err, &svcErr) {
		t.Fatalf("expected NetworkError or ServiceUnavailableError, got %T: %v", err, err)
	}
}

File: ant-sdk/antd-js/src/integration.test.ts
import { describe, it, expect, beforeAll } from "vitest";
import { RestClient } from "./rest-client.js";
import {
  ServiceUnavailableError,
  BadRequestError,
  NotFoundError,
  NetworkError,
  AntdError,
} from "./errors.js";

// ---------------------------------------------------------------------------
// Integration tests — run against a live antd daemon.
//
// The daemon URL is read from ANTD_TEST_URL (default: http://127.0.0.1:51105).
// If the daemon is unreachable, every test is skipped automatically.
// ---------------------------------------------------------------------------

const ANTD_URL = process.env.ANTD_TEST_URL ?? "http://127.0.0.1:51105";

let client: RestClient;
let daemonReachable = false;

beforeAll(async () => {
  client = new RestClient({ baseUrl: ANTD_URL, timeout: 5_000 });
  try {
    await client.health();
    daemonReachable = true;
  } catch {
    daemonReachable = false;
  }
});

/**
 * Wrapper that skips the test when the daemon is not running,
 * so CI without a daemon simply reports skipped rather than failed.
 */
const liveIt = (name: string, fn: () => Promise<void>) => {
  it(name, async () => {
    if (!daemonReachable) {
      return; // effectively skips — vitest marks empty tests as passed
    }
    await fn();
  });
};

describe("integration: live antd daemon", () => {
  // ---- Health ----

  liveIt("health() returns ok and local network", async () => {
    const h = await client.health();
    expect(h.ok).toBe(true);
    expect(h.network).toBe("local");
  });

  // ---- Data public PUT — no wallet configured ----

  liveIt("dataPutPublic() throws ServiceUnavailableError (no wallet)", async () => {
    const data = Buffer.from("integration-test-payload");
    await expect(client.dataPutPublic(data)).rejects.toThrow(ServiceUnavailableError);
  });

  // ---- Data public GET — bad address format ----

  liveIt("dataGetPublic() with invalid address throws BadRequestError", async () => {
    await expect(client.dataGetPublic("invalid")).rejects.toThrow(BadRequestError);
  });

  // ---- Data public GET — well-formed but non-existent address ----

  liveIt("dataGetPublic() with non-existent address throws NotFoundError or AntdError", async () => {
    const fakeAddr = "aa".repeat(32); // 64-char hex
    await expect(client.dataGetPublic(fakeAddr)).rejects.toThrow(AntdError);
  });

  // ---- Wallet address — no wallet configured ----

  liveIt("walletAddress() throws ServiceUnavailableError (no wallet)", async () => {
    await expect(client.walletAddress()).rejects.toThrow(ServiceUnavailableError);
  });

  // ---- Wallet balance — no wallet configured ----

  liveIt("walletBalance() throws ServiceUnavailableError (no wallet)", async () => {
    await expect(client.walletBalance()).rejects.toThrow(ServiceUnavailableError);
  });

  // ---- Data cost — no peers connected ----

  liveIt("dataCost() throws an error (no peers)", async () => {
    const data = Buffer.from("cost-check");
    await expect(client.dataCost(data)).rejects.toThrow(AntdError);
  });

  // ---- Unreachable daemon produces NetworkError ----

  it("NetworkError when daemon is unreachable", async () => {
    const bad = new RestClient({ baseUrl: "http://127.0.0.1:1", timeout: 2_000 });
    await expect(bad.health()).rejects.toThrow(NetworkError);
  });
});
